src/mono/mono/utils/mono-threads-posix.c

```diff
@@ -204,10 +213,25 @@ mono_threads_pthread_kill (MonoThreadInfo *info, int signum)
 	    result != ESRCH
 #if defined (__MACH__) && defined (ENOTSUP)
 	    && result != ENOTSUP
+#endif
+#if defined (__linux__)
+	    && !(result == EAGAIN && retry_count < signal_queue_ovf_retry_count)
 #endif
 	    )
 		g_error ("%s: pthread_kill failed with error %d - potential kernel OOM or signal queue overflow", __func__, result);
 
+#if defined (__linux__)
+	if (result == EAGAIN && retry_count < signal_queue_ovf_retry_count) {
+		/* HACK: if the signal queue overflows on linux, try again a couple of times.
+		 * Tries to address https://github.com/dotnet/runtime/issues/32377
+		 */
+		g_warning ("%s: pthread_kill failed with error %d - potential kernel OOM or signal queue overflow, sleeping for %ld microseconds", __func__, result, signal_queue_ovf_sleep_us);
+		g_usleep (signal_queue_ovf_sleep_us);
+		++retry_count;
+		goto redo;
+	}
```

Comment:
Won't you need to reset retry_count, otherwise this will stop only the first sequence of EGAINs?